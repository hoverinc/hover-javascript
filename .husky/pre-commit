#!/bin/sh
. "$(dirname "$0")/_/husky.sh"

node src pre-commit
