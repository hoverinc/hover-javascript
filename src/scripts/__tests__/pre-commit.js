import path from 'path'
import cases from 'jest-in-case'
import {
  unquoteSerializer,
  winPathSerializer,
} from '../../test/helpers/serializers'

expect.addSnapshotSerializer(unquoteSerializer)
expect.addSnapshotSerializer(winPathSerializer)

jest.mock('os', () => ({
  ...jest.requireActual('os'),
  tmpdir: jest.fn(() => '.test-tmp'),
}))

jest.mock('fs', () => ({
  ...jest.requireActual('fs'),
  mkdtempSync: jest.fn(prefix => `${prefix}TMPSUFFIX`),
  rmdirSync: jest.fn(),
  writeFileSync: jest.fn(),
}))

cases(
  'pre-commit',
  ({
    args = [],
    utils = require('../../utils'),
    hasPkgProp = () => false,
    hasFile = () => false,
    ifScript = () => true,
    expectError = false,
  }) => {
    // beforeEach
    const {sync: crossSpawnSyncMock} = require('cross-spawn')
    const {
      writeFileSync: writeFileSyncMock,
      rmdirSync: rmdirSyncMock,
    } = require('fs')

    const originalArgv = process.argv
    const originalExit = process.exit
    Object.assign(utils, {
      hasPkgProp,
      hasFile,
      ifScript,
      resolveBin: (modName, {executable = modName} = {}) => executable,
    })
    process.exit = jest.fn()

    process.argv = ['node', '../pre-commit', ...args]

    try {
      // tests
      require('../pre-commit')
      const commands = crossSpawnSyncMock.mock.calls.map(
        call => `${call[0]} ${call[1].join(' ')}`,
      )
      expect(commands).toMatchSnapshot()

      // ↓ ⚠️ Specific tests for when a custom test  ↓
      // ↓ command is supplied or DocToc is disabled ↓

      if (args.includes('--no-toc') || args.includes('--test-command')) {
        // ensure we don't pass invalid arguments through to `lint-staged`
        expect(
          crossSpawnSyncMock.mock.calls.some(([_command, commandArgs]) =>
            commandArgs.some(
              a =>
                a === '--testCommand' ||
                a === '--test-command' ||
                a === '--no-toc' ||
                a === '--noToc',
            ),
          ),
        ).not.toBeTruthy()

        const [writeFileSyncCall] = writeFileSyncMock.mock.calls

        // Snapshot the config file we use with the custom test command
        expect(writeFileSyncCall).toMatchSnapshot()

        // Make sure we clean up the temporary config file
        expect(rmdirSyncMock).toHaveBeenCalledWith(
          path.dirname(writeFileSyncCall[0]),
          {recursive: true},
        )
      }

      // ↑ ⚠️ Specific tests for when a custom test  ↑
      // ↑ command is supplied or DocToc is disabled ↑
    } catch (error) {
      if (expectError) {
        expect(error).toMatchSnapshot()
      } else {
        throw error
      }
    } finally {
      // afterEach
      process.exit = originalExit
      process.argv = originalArgv

      writeFileSyncMock.mockClear()

      jest.resetModules()
    }
  },
  {
    'calls lint-staged CLI with default args': {},
    'does not use built-in config with --config': {
      args: ['--config', './custom-config.js'],
    },
    'does not use built-in config with .lintstagedrc file': {
      hasFile: filename => filename === '.lintstagedrc',
    },
    'does not use built-in config with lint-staged.config.js file': {
      hasFile: filename => filename === 'lint-staged.config.js',
    },
    'does not use built-in config with lint-staged pkg prop': {
      hasPkgProp: prop => prop === 'lint-staged',
    },
    'forwards args': {
      args: ['--verbose'],
    },
    [`does not run validate script if it's not defined`]: {
      ifScript: () => false,
    },
    'throws an error when `--config` and `--testCommand` are used together': {
      expectError: true,
      args: [
        '--config',
        'some-config.js',
        '--testCommand',
        '"yarn test:unit --findRelatedTests"',
      ],
    },
    'throws an error when invalid `--testCommand` is provided': {
      expectError: true,
      args: ['--testCommand', '--config', 'some-config.js'],
    },
    'overrides built-in test command with --testCommand': {
      args: ['--testCommand', 'yarn test:custom --findRelatedTests foo.js'],
    },
    'overrides built-in test command with --test-command': {
      args: ['--test-command', 'yarn test:custom --findRelatedTests foo.js'],
    },
    'overrides built-in test command with --testCommand and forwards args': {
      args: [
        '--verbose',
        '--testCommand',
        '"yarn test:custom --findRelatedTests foo.js"',
      ],
    },
    'overrides built-in test command with --test-command and forwards args': {
      args: [
        '--verbose',
        '--test-command',
        'yarn test:custom --findRelatedTests foo.js',
      ],
    },
    'disables DocToc, overrides built-in test command, and forwards args': {
      args: [
        '--verbose',
        '--no-toc',
        '--test-command',
        'yarn test:custom --findRelatedTests foo.js',
        '--some-other-arg',
      ],
    },
    'disables DocToc and forwards args': {
      args: ['--verbose', '--no-toc', '--some-other-arg'],
    },
  },
)
