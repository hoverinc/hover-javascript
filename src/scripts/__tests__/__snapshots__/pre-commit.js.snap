// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pre-commit calls lint-staged CLI with default args 1`] = `
Array [
  lint-staged --config ./src/config/lintstagedrc.js,
  npm run validate,
]
`;

exports[`pre-commit disables DocToc and forwards args 1`] = `
Array [
  lint-staged --config .test-tmp/hover-javascriptTMPSUFFIX/.lintstaged.json --verbose --some-other-arg,
  npm run validate,
]
`;

exports[`pre-commit disables DocToc and forwards args 2`] = `
Array [
  .test-tmp/hover-javascriptTMPSUFFIX/.lintstaged.json,
  {"*.+(js|jsx|json|json5|yml|yaml|css|less|scss|ts|tsx|md|graphql|mdx|vue)":["./src/index.js format"],"*.+(js|jsx|ts|tsx)":["./src/index.js lint"],"*.+(tsx|ts|jsx|js)":["./src/index.js test --findRelatedTests"]},
]
`;

exports[`pre-commit disables DocToc, overrides built-in test command, and forwards args 1`] = `
Array [
  lint-staged --config .test-tmp/hover-javascriptTMPSUFFIX/.lintstaged.json --verbose --some-other-arg,
  npm run validate,
]
`;

exports[`pre-commit disables DocToc, overrides built-in test command, and forwards args 2`] = `
Array [
  .test-tmp/hover-javascriptTMPSUFFIX/.lintstaged.json,
  {"*.+(js|jsx|json|json5|yml|yaml|css|less|scss|ts|tsx|md|graphql|mdx|vue)":["./src/index.js format"],"*.+(js|jsx|ts|tsx)":["./src/index.js lint"],"*.+(tsx|ts|jsx|js)":["yarn test:custom --findRelatedTests foo.js"]},
]
`;

exports[`pre-commit does not run validate script if it's not defined 1`] = `
Array [
  lint-staged --config ./src/config/lintstagedrc.js,
]
`;

exports[`pre-commit does not use built-in config with .lintstagedrc file 1`] = `
Array [
  lint-staged ,
  npm run validate,
]
`;

exports[`pre-commit does not use built-in config with --config 1`] = `
Array [
  lint-staged --config ./custom-config.js,
  npm run validate,
]
`;

exports[`pre-commit does not use built-in config with lint-staged pkg prop 1`] = `
Array [
  lint-staged ,
  npm run validate,
]
`;

exports[`pre-commit does not use built-in config with lint-staged.config.js file 1`] = `
Array [
  lint-staged ,
  npm run validate,
]
`;

exports[`pre-commit forwards args 1`] = `
Array [
  lint-staged --config ./src/config/lintstagedrc.js --verbose,
  npm run validate,
]
`;

exports[`pre-commit overrides built-in test command with --test-command 1`] = `
Array [
  lint-staged --config .test-tmp/hover-javascriptTMPSUFFIX/.lintstaged.json,
  npm run validate,
]
`;

exports[`pre-commit overrides built-in test command with --test-command 2`] = `
Array [
  .test-tmp/hover-javascriptTMPSUFFIX/.lintstaged.json,
  {"README.md":["doctoc --maxlevel 4 --notitle"],"*.+(js|jsx|json|json5|yml|yaml|css|less|scss|ts|tsx|md|graphql|mdx|vue)":["./src/index.js format"],"*.+(js|jsx|ts|tsx)":["./src/index.js lint"],"*.+(tsx|ts|jsx|js)":["yarn test:custom --findRelatedTests foo.js"]},
]
`;

exports[`pre-commit overrides built-in test command with --test-command and forwards args 1`] = `
Array [
  lint-staged --config .test-tmp/hover-javascriptTMPSUFFIX/.lintstaged.json --verbose,
  npm run validate,
]
`;

exports[`pre-commit overrides built-in test command with --test-command and forwards args 2`] = `
Array [
  .test-tmp/hover-javascriptTMPSUFFIX/.lintstaged.json,
  {"README.md":["doctoc --maxlevel 4 --notitle"],"*.+(js|jsx|json|json5|yml|yaml|css|less|scss|ts|tsx|md|graphql|mdx|vue)":["./src/index.js format"],"*.+(js|jsx|ts|tsx)":["./src/index.js lint"],"*.+(tsx|ts|jsx|js)":["yarn test:custom --findRelatedTests foo.js"]},
]
`;

exports[`pre-commit overrides built-in test command with --testCommand 1`] = `
Array [
  lint-staged --config .test-tmp/hover-javascriptTMPSUFFIX/.lintstaged.json,
  npm run validate,
]
`;

exports[`pre-commit overrides built-in test command with --testCommand and forwards args 1`] = `
Array [
  lint-staged --config .test-tmp/hover-javascriptTMPSUFFIX/.lintstaged.json --verbose,
  npm run validate,
]
`;

exports[`pre-commit throws an error when \`--config\` and \`--testCommand\` are used together 1`] = `[Error: @hover/javascript/pre-commit: --config and --testCommand cannot be used together (--testCommand only works with built-in lint-staged configuration)]`;

exports[`pre-commit throws an error when invalid \`--testCommand\` is provided 1`] = `[Error: @hover/javascript/pre-commit: invalid --testCommand]`;
